gprofiler/py_package_version.py

```diff
@@ -0,0 +1,160 @@
+        return None
+    # This extra Path-str cast normalizes entries.
+    return (str(pathlib.Path(row[0])) for row in csv.reader(text.splitlines()))
+
+
+def _files_from_legacy(dist: pkg_resources.Distribution) -> Optional[Iterator[str]]:
+    try:
+        text = dist.get_metadata("installed-files.txt")
+    except:
+        return None
+    paths = (p for p in text.splitlines(keepends=False) if p)
+    root = dist.location
+    info = dist.egg_info
+    if root is None or info is None:
+        return paths
+    try:
+        info_rel = pathlib.Path(info).relative_to(root)
+    except ValueError:  # info is not relative to root.
+        return paths
+    if not info_rel.parts:  # info *is* root.
+        return paths
+    return (_convert_legacy_entry(pathlib.Path(p).parts, info_rel.parts) for p in paths)
+
+
+def _get_package_name(dist: pkg_resources.Distribution) -> Optional(str):
+    # TODO: Test
+    metadata = _get_metadata(dist)
+    if metadata:
+        # The metadata should NEVER be missing the Name: key, but if it somehow
+        # does, fall back to the known canonical name.
+        return metadata.get("Name", dist.project_name)
+    return None
+
+
+def get_versions(modules_paths: List[str], pid: int):
+    """Return a dict with module_path: (package_name, version). If couldn't
+    determine the version the value is None.
+
+    modules_paths must be absoulte. pid is required to access the path via
+    /proc/{pid}/root/.
+
+    Given a path to a module, it's not trivial to determine its package as some
+    packages contain modules with completely unrelated names (e.g. the module
+    pkg_resources used in this function is a part of the setuptools package).
+    This function gathers info for all the packages in the (dist|site)-packages
+    directory in each path, including a list of all the files associated with
+    each package. This list is searched for the given module path.
+    """
+    result = dict.fromkeys(modules_paths)
+    path_to_dist = {}
+
+    for path in modules_paths:
+        if not path.startswith("/"):
+            continue
+
+        if path not in path_to_dist:
+            packages_path = _get_packages_dir(path)
+            if packages_path is None:
+                # This module is (probably) not part of a package
+                continue
+            packages_path = _convert_to_proc_root_path(packages_path, pid)
+
+            for dist in pkg_resources.find_distributions(packages_path):
+                files_iter = _files_from_record(dist) or _files_from_legacy(dist)
+                if files_iter is not None:
+                    path_to_dist.update(dict.fromkeys((os.path.join(packages_path, file) for file in files_iter), dist))
+
+        dist = path_to_dist.get(_convert_to_proc_root_path(path, pid))
+        if dist is not None:
+            name = _get_package_name(dist)
+            if name is not None:
+                result[path] = (_get_package_name(dist), dist.version)
+
```

Comment:
Maybe just store package name, version tuple in `path_to_dist` dict values, because now you're parsing the metadata file on each call to `_get_package_name` which does IO that can be reduced easily

Reply:
I'm not sure it's true. I might use `_get_package_name` twice for the same dist, but on the other hand I perform it only for the dists of the relevant modules as opposed to all the many dists in the directory.